pkg/controller/controller_utils.go

```diff
@@ -0,0 +1,281 @@
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client/record"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util/wait"
+	"github.com/golang/glog"
+	"sync/atomic"
+	"time"
+)
+
+const CreatedByAnnotation = "kubernetes.io/created-by"
+
+// Expectations are a way for replication controllers to tell the rc manager what they expect. eg:
+//	RCExpectations: {
+//		rc1: expects  2 adds in 2 minutes
+//		rc2: expects  2 dels in 2 minutes
+//		rc3: expects -1 adds in 2 minutes => rc3's expectations have already been met
+//	}
+//
+// Implementation:
+//	PodExpectation = pair of atomic counters to track pod creation/deletion
+//	RCExpectationsStore = TTLStore + a PodExpectation per rc
+//
+// * Once set expectations can only be lowered
+// * An RC isn't synced till its expectations are either fulfilled, or expire
+// * Rcs that don't set expectations will get woken up for every matching pod
+
+// expKeyFunc to parse out the key from a PodExpectation
+var expKeyFunc = func(obj interface{}) (string, error) {
+	if e, ok := obj.(*PodExpectations); ok {
+		return e.key, nil
+	}
+	return "", fmt.Errorf("Could not find key for obj %#v", obj)
+}
+
+// RCExpectations is a ttl cache mapping rcs to what they expect to see before being woken up for a sync.
+type RCExpectations struct {
+	cache.Store
+}
+
+// GetExpectations returns the PodExpectations of the given rc.
+func (r *RCExpectations) GetExpectations(rc *api.ReplicationController) (*PodExpectations, bool, error) {
+	rcKey, err := rcKeyFunc(rc)
+	if err != nil {
+		return nil, false, err
+	}
+	if podExp, exists, err := r.GetByKey(rcKey); err == nil && exists {
+		return podExp.(*PodExpectations), true, nil
+	} else {
+		return nil, false, err
+	}
+}
+
+// SatisfiedExpectations returns true if the replication manager has observed the required adds/dels
+// for the given rc. Add/del counts are established by the rc at sync time, and updated as pods
+// are observed by the replication manager's podController.
+func (r *RCExpectations) SatisfiedExpectations(rc *api.ReplicationController) bool {
+	if podExp, exists, err := r.GetExpectations(rc); exists {
+		if podExp.Fulfilled() {
+			return true
+		} else {
+			glog.V(4).Infof("Controller %v still waiting on expectations %#v", podExp)
+			return false
+		}
+	} else if err != nil {
+		glog.V(2).Infof("Error encountered while checking expectations %#v, forcing sync", err)
+	}
+	// Trigger a sync if we either encountered and error (which shouldn't happen since we're
+	// getting from local store) or this rc hasn't established expectations.
+	return true
+}
+
+// setExpectations registers new expectations for the given rc. Forgets existing expectations.
+func (r *RCExpectations) setExpectations(rc *api.ReplicationController, add, del int) error {
+	rcKey, err := rcKeyFunc(rc)
+	if err != nil {
+		return err
+	}
+	return r.Add(&PodExpectations{add: int64(add), del: int64(del), key: rcKey})
```

Comment:
Oh, I see, every set blows away whatever was in there previously. I guess that makes sense.